description: Add option to enable desktop mode all the time
author: Wengling Chen <feiyu2817@gmail.com>

---
 chrome/android/java/res/xml/accessibility_preferences.xml                                                     |    5 ++++
 chrome/android/java/src/org/chromium/chrome/browser/accessibility/settings/AccessibilitySettings.java         |   10 +++++++++
 chrome/android/java/src/org/chromium/chrome/browser/tabmodel/TabModelSelectorImpl.java                        |   11 ++++++++++
 chrome/browser/preferences/android/java/src/org/chromium/chrome/browser/preferences/ChromePreferenceKeys.java |    3 ++
 chrome/browser/ui/android/strings/android_chrome_strings.grd                                                  |    6 +++++
 5 files changed, 35 insertions(+)
--- a/chrome/android/java/res/xml/accessibility_preferences.xml
+++ b/chrome/android/java/res/xml/accessibility_preferences.xml
@@ -25,6 +25,11 @@
         android:summary="@string/accessibility_tab_switcher_summary"
         android:title="@string/accessibility_tab_switcher_title" />
 
+    <org.chromium.components.browser_ui.settings.ChromeBaseCheckBoxPreference
+        android:key="force_desktop_mode"
+        android:summary="@string/force_desktop_mode_summary"
+        android:title="@string/force_desktop_mode_title" />
+
     <Preference
         android:key="captions"
         android:title="@string/accessibility_captions_title"/>
--- a/chrome/browser/ui/android/strings/android_chrome_strings.grd
+++ b/chrome/browser/ui/android/strings/android_chrome_strings.grd
@@ -1204,6 +1204,12 @@ Your Google account may have other forms
       <message name="IDS_ACCESSIBILITY_CAPTIONS_TITLE" desc="Title of the preference that allows the user to update caption settings.">
         Captions
       </message>
+      <message name="IDS_FORCE_DESKTOP_MODE_SUMMARY" desc="Summary of the preference that allows the user to force all pages to be opened in desktop mode.">
+       Open all pages in desktop mode
+      </message>
+      <message name="IDS_FORCE_DESKTOP_MODE_TITLE" desc="Title of the preference that allows the user to update force desktop mode settings.">
+        Force Desktop Mode
+      </message>
 
       <!-- Language Settings -->
       <message name="IDS_LANGUAGE_SETTINGS" desc="Title for the Languages settings screen that allows users to manage their languages preferences. [CHAR_LIMIT=32]">
--- a/chrome/browser/preferences/android/java/src/org/chromium/chrome/browser/preferences/ChromePreferenceKeys.java
+++ b/chrome/browser/preferences/android/java/src/org/chromium/chrome/browser/preferences/ChromePreferenceKeys.java
@@ -503,6 +503,8 @@ public final class ChromePreferenceKeys
     public static final String FONT_USER_FONT_SCALE_FACTOR = "user_font_scale_factor";
     public static final String FONT_USER_SET_FORCE_ENABLE_ZOOM = "user_set_force_enable_zoom";
 
+    public static final String FLAGS_FORCE_DESKTOP_MODE_ENABLED = "Chrome.UngoogledChromium.ForceDesktopMode";
+
     public static final String HISTORY_SHOW_HISTORY_INFO = "history_home_show_info";
 
     /** Keys used to save settings related to homepage. */
@@ -1041,6 +1043,7 @@ public final class ChromePreferenceKeys
                 FIRST_RUN_SKIPPED_BY_POLICY,
                 FLAGS_CACHED.pattern(),
                 FLAGS_FIELD_TRIAL_PARAM_CACHED.pattern(),
+                FLAGS_FORCE_DESKTOP_MODE_ENABLED,
                 FLAGS_LAST_CACHED_MINIMAL_BROWSER_FLAGS_TIME_MILLIS,
                 HOMEPAGE_LOCATION_POLICY,
                 HOMEPAGE_USE_CHROME_NTP,
--- a/chrome/android/java/src/org/chromium/chrome/browser/accessibility/settings/AccessibilitySettings.java
+++ b/chrome/android/java/src/org/chromium/chrome/browser/accessibility/settings/AccessibilitySettings.java
@@ -34,6 +34,7 @@ public class AccessibilitySettings
     static final String PREF_READER_FOR_ACCESSIBILITY = "reader_for_accessibility";
     static final String PREF_CAPTIONS = "captions";
     static final String PREF_IMAGE_DESCRIPTIONS = "image_descriptions";
+    static final String PREF_FORCE_DESKTOP_MODE = "force_desktop_mode";
 
     private TextScalePreference mTextScalePref;
     private ChromeBaseCheckBoxPreference mForceEnableZoomPref;
@@ -80,6 +81,12 @@ public class AccessibilitySettings
                                                       .getBoolean(Pref.READER_FOR_ACCESSIBILITY));
         readerForAccessibilityPref.setOnPreferenceChangeListener(this);
 
+        ChromeBaseCheckBoxPreference forceDesktopModePref =
+                (ChromeBaseCheckBoxPreference) findPreference(PREF_FORCE_DESKTOP_MODE);
+        forceDesktopModePref.setChecked(SharedPreferencesManager.getInstance().readBoolean(
+                            ChromePreferenceKeys.FLAGS_FORCE_DESKTOP_MODE_ENABLED, false));
+        forceDesktopModePref.setOnPreferenceChangeListener(this);
+
         ChromeBaseCheckBoxPreference mAccessibilityTabSwitcherPref =
                 (ChromeBaseCheckBoxPreference) findPreference(
                         ChromePreferenceKeys.ACCESSIBILITY_TAB_SWITCHER);
@@ -134,6 +141,9 @@ public class AccessibilitySettings
         } else if (PREF_READER_FOR_ACCESSIBILITY.equals(preference.getKey())) {
             UserPrefs.get(Profile.getLastUsedRegularProfile())
                     .setBoolean(Pref.READER_FOR_ACCESSIBILITY, (Boolean) newValue);
+        } else if (PREF_FORCE_DESKTOP_MODE.equals(preference.getKey())) {
+            SharedPreferencesManager.getInstance().writeBoolean(
+                ChromePreferenceKeys.FLAGS_FORCE_DESKTOP_MODE_ENABLED, (Boolean) newValue);
         }
         return true;
     }
--- a/chrome/android/java/src/org/chromium/chrome/browser/tabmodel/TabModelSelectorImpl.java
+++ b/chrome/android/java/src/org/chromium/chrome/browser/tabmodel/TabModelSelectorImpl.java
@@ -11,6 +11,8 @@ import androidx.annotation.VisibleForTes
 
 import org.chromium.base.supplier.Supplier;
 import org.chromium.chrome.browser.compositor.layouts.content.TabContentManager;
+import org.chromium.chrome.browser.preferences.ChromePreferenceKeys;
+import org.chromium.chrome.browser.preferences.SharedPreferencesManager;
 import org.chromium.chrome.browser.flags.ActivityType;
 import org.chromium.chrome.browser.profiles.Profile;
 import org.chromium.chrome.browser.tab.SadTab;
@@ -162,6 +164,15 @@ public class TabModelSelectorImpl extend
 
             @Override
             public void onPageLoadFinished(Tab tab, GURL url) {
+                if (SharedPreferencesManager.getInstance().readBoolean(
+                    ChromePreferenceKeys.FLAGS_FORCE_DESKTOP_MODE_ENABLED, false)) {
+                    final boolean reloadOnChange = !tab.isNativePage();
+                    final boolean usingDesktopUserAgent =
+                        tab.getWebContents().getNavigationController().getUseDesktopUserAgent();
+                    if (!usingDesktopUserAgent)
+                        tab.getWebContents().getNavigationController().setUseDesktopUserAgent(
+                            !usingDesktopUserAgent, reloadOnChange);
+                }
                 tab.getId();
             }
 
